samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -388,106 +415,77 @@ public ContainerStorageManager(
   }
 
   /**
-   * Create taskStores for all stores in storageEngineFactories.
-   * The store mode is chosen as bulk-load if its a non-sideinput store, and readWrite if its a sideInput store
+   * Create taskStores for all stores in storesToCreate.
+   * The store mode is chosen as read-write mode.
    */
-  private Map<TaskName, Map<String, StorageEngine>> createTaskStores(ContainerModel containerModel, JobContext jobContext, ContainerContext containerContext,
+  private Map<TaskName, Map<String, StorageEngine>> createTaskStores(Set<String> storesToCreate,
+      ContainerModel containerModel, JobContext jobContext, ContainerContext containerContext,
       Map<String, StorageEngineFactory<Object, Object>> storageEngineFactories, Map<String, Serde<Object>> serdes,
       Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics,
       Map<TaskName, TaskInstanceCollector> taskInstanceCollectors) {
-
     Map<TaskName, Map<String, StorageEngine>> taskStores = new HashMap<>();
+    StorageConfig storageConfig = new StorageConfig(config);
 
-    // iterate over each task in the containerModel, and each store in storageEngineFactories
+    // iterate over each task and each storeName
     for (Map.Entry<TaskName, TaskModel> task : containerModel.getTasks().entrySet()) {
       TaskName taskName = task.getKey();
       TaskModel taskModel = task.getValue();
-
       if (!taskStores.containsKey(taskName)) {
         taskStores.put(taskName, new HashMap<>());
       }
 
-      for (String storeName : storageEngineFactories.keySet()) {
-
-        StorageEngineFactory.StoreMode storeMode = this.taskSideInputStoreSSPs.get(taskName).containsKey(storeName) ?
-            StorageEngineFactory.StoreMode.ReadWrite : StorageEngineFactory.StoreMode.BulkLoad;
+      for (String storeName : storesToCreate) {
+        // A store is considered durable if it is backed by a changelog or another backupManager factory
+        boolean isDurable = changelogSystemStreams.containsKey(storeName) ||
+            !storageConfig.getStoreBackupManagerClassName(storeName).isEmpty();
+        boolean isSideInput = this.taskSideInputStoreSSPs.get(taskName).containsKey(storeName);
```

Comment:
can use `sideInputStores` set instead to check if a store is side input store instead of two look ups.